### ISSUE TITLE
Latency modeling for sandbox/real-time

### ISSUE THREAD

**Issue by cjdsellers** [nautechsystems/nautilus_trader#1677]
# Feature Request

A `use_message_queue` config option was added to `SimulatedExchange` as a temporary work around for processing trading commands in real-time cbfc7035aa62df6cdf6c45cf3621cf3d0eea4faf to accommodate the "sandbox mode".

This means that currently when trading commands are submitted to the sandbox execution client (and onto its internal simulated exchange) there is practically no latency in either direction. The commands will be processed and events generated at local platform latency (microseconds) which is relatively unrealistic for most venue and setups.

A future feature could leverage the `LatencyModel` to extend to real-time, potentially by using a timer to "delay" command processing and events being generated - to achieve to desired latency model spec.

**Comment by TheLegendAli** [comment]
yup we need something life that, for my backtest I'm implementing few latencies:

- latency that I receive the data

- latency for nautilus to do its magic

- latency for exchange to receive order from nautilus

I'm manual using "timer" to add delay. But would be nice if we can automate it

**Comment by ChrisAB** [comment]
Was wondering if this is still needed and if I could pick it up? @TheLegendAli

**Comment by TheLegendAli** [comment]
i think so. I would use it everyday :)

**Comment by ChrisAB** [comment]
According to https://github.com/nkaz001/hftbacktest/blob/master/docs/latency_models.rst
This seems like two PRs instead of 1?
My question is, currently, response latency is missing in latency model. Should this be added? Should this also take base_latency into account or no? Cause if it does, I think it will break existing configurations as it will add base_latency into every call.
Or should this issue only focus on actually adding latency model to sandbox?

**Comment by TheLegendAli** [comment]
Hi Chris,

Sorry for the late reply, I was away for a bit. The latency model is everything I asked for and then some — really nice work.

To your questions:

**Response latency:** I don't need it personally, but I'd assume others will, so worth including.

**base_latency:** my instinct is that it's a one-way figure, which would mean it applies to both legs. If I read you correctly, this will break the current config? If so, would it be cleaner to just split it into two explicit parameters?

1. a `base latency` for sending orders
2. a separate `response_latency` for the fill coming back

**Sandbox only:** yes, i would imagine we only need it in sandbox.

Let me know if I answered your questions

Thanks again for picking this up.

**Comment by ChrisAB** [comment]
No problem, I was also away for quite a bit. I also took a look at the discussion created by vedjaw.

I actually implemented both parts. adding response_latency was the easy part. However, I cannot say the same about sandbox latency.

I will definitely need some advice from @cjdsellers as the PR would be ~3000 lines total. (Around 70% of that are tests)
This implementation covers **commands only**, as venue-generated events (accepts, fills, cancels, expirations) still dispatch synchronously. Default is no latency model.

What I implemented:
1. `SandboxExecutionClientConfig` gains `latency_model` which can take `StaticLatencyModel` and mirrors the current implementation of `fee_model`.
```python
config = SandboxExecutionClientConfig(
    venue=Venue("BINANCE"),
    starting_balances=[Money.from_str("10_000 USDT")],
    latency_model=StaticLatencyModel(base_latency_nanos=1_000_000_000),
)
```
2. Commands are pushed into a `BinaryHeap` in `SandboxInner` instead of hitting the matching engine directly. (queries are not delayed as they are local cache reads)
3. submit and submit-list use the insert leg, modify and batch-modify the update leg, cancel / batch-cancel / cancel-all the delete leg.
4. **Due time is keyed off arrival at the client** (`now + leg`), not `command.ts_init()`. This is different than backtest as the two can differ by real wall-clock time spent upstream.
5. FIFO within a due timestamp uses a single monotonic sequence rather than the backtest's per-due-timestamp counter map. The backtest wipes a timestamp's counter as it drains it (exchange.rs:1463), which is safe there because its drain loop is exhaustive, nothing is left at that timestamp to sort against. My drain has a second exit (the hold-back in point 8), so it can leave a command queued at a timestamp it has already popped from; a later command landing on that same timestamp would then restart the count at 1 and sort ahead of it. A sequence that only increments has no reset to exploit.
6. The drain is timer-driven from a `LiveClock` alert armed at the earliest queued due time, meaning a command lands even if no further market data arrives.
7. Immediate and deferred paths share one code path. I moved the command method from `ExecutionClient` into `apply_*` helpers in `SandboxInner`. The method chooses "apply now" or "enqueue". This is a lot of the change for `execution.rs` and where a lot of lines of code come from, and it is technically moved logic rather than new logic.
8. A bit of a bigger point. There were two problems I encountered that I had to solve:
  A) A command can overtake the submit that would have created its matching engine (engines are created lazily on first submit, unlike backtest) Left alone, a deferred cancel finds no engine and is discarded with no event (the order stays `PENDING_CANCEL` forever, since the sandbox emits no order status reports). So the deferred path builds the engine from the cached instrument, and when  the instrument itself is absent it rejects it directly.
  B) Two commands for the same order falling due in the same drain pass. Concretely, `process_cancel_all` reads `cache.orders_open(...)` and would miss an order submitted earlier in the same pass. The drain therefore applies at most one command per order per pass and retries the rest ~1ms later, with a floor so a later enqueue cannot re-arm the alert onto the held-back command's elapsed due time.
9. Teardown. `stop()` cancels the alert and discards anything in flight, rejecting each command (`OrderRejected` / `OrderModifyRejected` / `OrderCancelRejected`) so no order is stranded. A discarded `CancelAllOrders` is dropped silently instead: the strategy marks no order `PENDING_CANCEL` for it, and the order FSM has only one `CancelRejected` transition, `(PendingCancel, CancelRejected)`, so there is no valid rejection to synthesize. `reset()` clears the queue without rejecting.
10. Again, a change compared to backtest. Backtest `process_modify_submitted_order` amends an order still in flight when a modify overtakes its submit. I did **not** port that: here the modify goes to the venue and is rejected, as a real venue would for an order it has not received. 

For the exact lines of code needed in execution.rs (+700 lines):
1. apply_* extraction, venue-side bodies moved out of the 7 ExecutionClient methods
2. Queue + timer: DelayedCommand, heap ordering, enqueue, command_leg_latency, drain_inbound_due, arm_inbound_alert{,_at}, drain_due_now
3. Out-of-order arrival: apply_trading_command, ensure_engine_for, reject_command/_modify/_cancel
4. Same-pass hold-back: AppliedThisPass, conflicts_with, record, for_each_order_id
5. Teardown/reset: discard_inbound_queue, clear_inbound_queue, stop/reset hooks
6. Wiring: 6 new SandboxInner fields (39), Rc::new_cyclic construction (17), 7 call-site if/else branches (30), 10 drain_due_now() hooks (20), imports/const (30)

A few more for configs, serde guards, getters etc. (About 100 lines)
Then about 2000 lines of tests.

I have some questions regarding these points:
1. Do you want this at all right now, given it is on the backlog? I know v2 migration is going on, I can also just create the PR, or just park current implementation all-together now, or deliver just the response latency since that is a bit more straight forward.
2. Point 10, do you wanna keep it realistic as I implemented it now, or do you want it the same as backtest?
3. Is the one-command-per-order-per-pass approach in point 8 reasonable, or would you rather the sandbox not batch a drain at all?
4. Should this be split? I am not sure exactly how to split it in this exact moment, but I can figure something out.
5. Just to be 100% sure, is it okay if I first deliver this, and then in the second half I deliver the response latency on both sandbox and backtest? (Or, alternatively, I can keep backtest and sandbox response latency on separate PRs, probably easier to review aswell)

**Comment by TheLegendAli** [comment]
Thanks for going this deep.

1. I'd split `response_latency` off. More modular, and it opens the smaller piece up to feedback without gating it on the larger review.

2. Keep your version — I think the call-out is right.

3. I ran into this same problem myself for a different reason, and landed on roughly your approach, just implemented a bit differently. I think it's the right call here.

4. That's for you and @cjdsellers  to decide — I won't get in the way of the master chefs who are cooking.

5. Absolutely.

**Comment by graceyangfan** [comment]
Hi @cjdsellers — I would like to validate a Lighter-specific backtesting use case for this issue.

Lighter account tiers expose different latency paths for maker/taker and cancel/modify operations. The key modeling problem is that a regular LIMIT can be marketable, partially take liquidity, and leave a residual resting order. Therefore the venue-specific rule for choosing the pre-arrival latency lane should not be hard-coded into Nautilus or inferred from the final `LiquiditySide`.

Would a pluggable `OrderLatencyModel`, analogous to `FeeModel`, be appropriate? The engine would pass a read-only submission context (order, command type, venue/account metadata, and optionally the current book); the user/venue model would return insert, modify, cancel, and response latencies. The MatchingEngine would remain responsible for actual matching and final maker/taker classification after arrival.

Desired semantics:
- submit: model returns an arrival delay; the order cannot fill before arrival;
- modify/cancel: independent delays; the old order remains live until the command arrives;
- response latency is separate;
- `SubmitOrderList` can use either per-order or list-level policy;
- existing `StaticLatencyModel` remains backward compatible.

This keeps Lighter-specific semantics in a user-provided model while preserving Nautilus’ generic matching core. References: [Lighter account tiers](https://apidocs.lighter.xyz/docs/account-types), [Lighter matching rules](https://docs.lighter.xyz/trading/orders-and-matching), and [hftbacktest latency model patterns](https://hft.readthedocs.io/en/latest/latency_models.html).

Does this fit the direction of #1677, or would you prefer a different abstraction? A small Rust prototype with lifecycle tests could validate the API.
